httpql/src/main/java/com/hubspot/httpql/internal/CombinedFilterEntry.java

```diff
@@ -0,0 +1,107 @@
+package com.hubspot.httpql.internal;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import org.jooq.Condition;
+import org.jooq.Operator;
+import org.jooq.impl.DSL;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Lists;
+import com.hubspot.httpql.FieldFactory;
+import com.hubspot.httpql.QuerySpec;
+
+public class CombinedFilterEntry<T extends QuerySpec> implements FilterEntryConditionCreator<T> {
+
+  private Operator operator;
+
+  private List<FilterEntryConditionCreator<T>> conditionCreators;
+
+  public CombinedFilterEntry(Operator operator, List<FilterEntryConditionCreator<T>> conditionCreators) {
+    this.operator = operator;
+    this.conditionCreators = Lists.newArrayList(conditionCreators);
+  }
+
+  @Override
+  public Condition getCondition(QuerySpec value, FieldFactory fieldFactory) {
+    return DSL.condition(operator, conditionCreators.stream()
+        .map(cc -> cc.getCondition(value, fieldFactory))
+        .collect(Collectors.toList()));
+  }
+
+  @Override
+  public Collection<BoundFilterEntry<T>> getFlattenedBoundFilterEntries() {
+    return conditionCreators.stream()
+        .map(FilterEntryConditionCreator::getFlattenedBoundFilterEntries)
+        .flatMap(Collection::stream)
+        .collect(ImmutableList.toImmutableList());
+  }
+
+  public void addConditionCreator(FilterEntryConditionCreator<T> conditionCreator) {
+    if (!getFlattenedBoundFilterEntries().contains(conditionCreator)) {
+      conditionCreators.add(conditionCreator);
+    }
+  }
+
+  public boolean removeAllFiltersFor(String fieldName) {
+    boolean removedFromThis = conditionCreators.removeIf(f -> f instanceof BoundFilterEntry && ((BoundFilterEntry<T>) f)
+        .getQueryName()
+        .equals(fieldName));
+    boolean removedFromChildren = conditionCreators.stream()
+        .filter(cc -> cc instanceof CombinedFilterEntry)
+        .map(cc -> ((CombinedFilterEntry<T>) cc).removeAllFiltersFor(fieldName))
+        .reduce(Boolean::logicalOr).orElse(false);
+
+    return removedFromThis || removedFromChildren;
+  }
+
+  public Optional<BoundFilterEntry<T>> getFirstFilterForFieldName(String fieldName) {
```

Comment:
I'm not sure I totally understand what a "first" filter means in this kind of tree structure, or what it's used for.  It looks like it'll first look through all top-level "leaf" filters and then dive into the combined filters... Will that find the "first"?

What about using `getFlattenedBoundFilterEntries` for this?

Reply:
Yea this is a part of the API that was a little odd to maintain in the new structure. I think `getFlattenedBoundFilterEntries` should be fine actually.